controllers/configmap_controller.go

```diff
@@ -18,45 +18,229 @@ package controllers
 
-	// your logic here
+	var err error
+	// Create a new signer using the private key that the instances will be configured with
+	r.signer, err = secrets.NewSignerFromSecret(kubeTypes.NamespacedName{Namespace: r.watchNamespace,
+		Name: secrets.PrivateKeySecret}, r.client)
+	if err != nil {
+		if k8sapierrors.IsNotFound(err) {
+			return ctrl.Result{}, errors.Wrapf(err, "%s does not exist, please create it", secrets.PrivateKeySecret)
+		}
+		return ctrl.Result{}, errors.Wrapf(err, "unable to get signer from secret %s", req.NamespacedName)
+	}
+
+	// Fetch the ConfigMap. The predicate will have filtered out any ConfigMaps that we should not reconcile
+	// so it is safe to assume that all ConfigMaps being reconciled describe hosts that need to be present in the
+	// cluster.
+	configMap := &core.ConfigMap{}
+	if err := r.client.Get(ctx, req.NamespacedName, configMap); err != nil {
+		if k8sapierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, r.addAndRemoveNodes(ctx, configMap)
+}
+
+// parseHosts gets the lists of hosts specified in the configmap's data
+func (r *ConfigMapReconciler) parseHosts(configMapData map[string]string) ([]*instanceInfo, error) {
+	hosts := make([]*instanceInfo, len(configMapData))
+	entryCount := 0
+	// Get information about the hosts from each entry. The expected key/value format for each entry is:
+	// <address>: username=<username>
+	for address, data := range configMapData {
+		if err := addressValid(address); err != nil {
+			return nil, errors.Wrapf(err, "invalid address %s", address)
+		}
+		splitData := strings.SplitN(data, "=", 2)
+		if len(splitData) == 0 || splitData[0] != "username" {
+			return hosts, fmt.Errorf("data for entry %s has an incorrect format", address)
+		}
+
+		hosts[entryCount] = newInstanceInfo(address, splitData[1])
+		entryCount++
+	}
+	return hosts, nil
+}
+
+// addressValid checks that the given address is either an ipv4 address, or resolves to any ip address
+func addressValid(address string) error {
+	// first check if address is an IP address
+	if parsedAddr := net.ParseIP(address); parsedAddr != nil {
+		if parsedAddr.To4() != nil {
+			return nil
+		}
+		// if the address parses into an IP but is not ipv4 it must be ipv6
+		return fmt.Errorf("ipv6 is not supported")
+	}
+	// Do a check that the DNS provided is valid
+	addressList, err := net.LookupHost(address)
+	if err != nil {
+		return errors.Wrapf(err, "error looking up DNS")
+	}
+	if len(addressList) == 0 {
+		return fmt.Errorf("DNS did not resolve to an address")
+	}
+	return nil
+}
+
+// addAndRemoveNodes corrects the discrepancy between the "expected" hosts slice, and the "actual" nodelist
```

Comment:
Won't `reconcileConfigMap` or `reconcileInstances` or `reconcileNodes` be more appropriate here? `addAndRemoveNodes` implies you are always doing both while you could being doing one or the other.